### DISCUSSION TITLE
Add string comparison for Update platform

### DISCUSSION THREAD

**Discussion by chemelli74** [home-assistant/architecture#1130] [Entity Models] [answered]
Current implementation of version comparison is based on AwesomeVersion:

https://github.com/home-assistant/core/blob/edad766fd3c30cd88676898e5219fff394e9c7ff/homeassistant/components/update/__init__.py#L182-L185

For some devices, ex. Shelly gen 1 devices, this always fails with `AwesomeVersionCompareException`.

Strings are as follow:

- latest_version: "`20230913-111730/v1.14.0-gcb84623`"
- installer_version: "`20231107-162609/v1.14.1-rc1-g0617c15`"

Note: Shelly gen2/3 devices work fine as the strings are "`2.1.0`" and "`2.2.0-beta7`".

Proposed change is to introduce the ability to make a pure string comparison.
Here the PR with the implementation: https://github.com/home-assistant/core/pull/124797

**Comment by chemelli74** [comment]
Another option could be to introduce 2 new methods that are optionally implemented by each integration:

- latest_version_parsing()
- installed_version_parsing()

and modify only the state method:

```python
latest_version_parsing = self.latest_version_parsing or self.latest_version
installed_version_parsing = self.installed_version_parsing or self.installed_version
 try:
            newer = _version_is_newer(latest_version_parsing, installed_version_parsing)
        except AwesomeVersionCompareException:
            # Can't compare versions, already tried exact match
            return STATE_ON
        return STATE_ON if newer else STATE_OFF
```

**Comment by emontnemery** [comment, marked as answer]
Another proposal: Make the `_version_is_newer` function a method of `UpdateEntity` so integrations can override it. Would that work @chemelli74 ?

**Reply by chemelli74** [reply]
Yes, I would say so.
